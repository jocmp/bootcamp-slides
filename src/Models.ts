export interface SlideshowModel {
    id: number;
    title: string;
    slides: SlideModel[];
}

export interface SlideModel {
    title: string;
    slide_type: "title" | "simple" | "two-column";
    content?: string[];
}

export interface AppState {
    slideshow: SlideshowModel;
    viewedIndices: number[];
    loading: boolean;
    error: string;
}

export class LinkFactory {

    static createSlideshowLink(id: number, index: number = 0) {
        return `/slideshows/${id}/slides/${index}`
    }
}