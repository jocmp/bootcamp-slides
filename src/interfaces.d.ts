declare module 'enzyme-to-json'