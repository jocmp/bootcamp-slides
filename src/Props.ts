import { SlideModel, SlideshowModel } from './Models'

export interface SlideshowOwnProps {
    history: any;
    match: { params: any };
}

export interface SlideshowProps extends SlideshowStateProps, SlideshowDispatchProps, SlideshowOwnProps { };

export interface SlideshowStateProps {
    slideshow: SlideshowModel;
    viewedIndices: number[];
}

export interface SlideshowDispatchProps {
    handleNextSlide: (id: string, index: string, length: number, history: any) => any;
    handlePreviousSlide: (id: string, index: string, length: number, history: any) => any;
    fetchSlideshow: (id: number, index: number) => void;
    viewSlide: (id: number) => any;
}

export interface OverviewProps {
    match: { params: any };
    slideshow: SlideshowModel;
    viewedIndices: number[];
    viewSlide: (id: number) => any;
}

export interface OverviewItemProps {
    slideshow_id: number;
    index: number;
    slide: SlideModel;
    viewed: boolean;
    isCurrent: boolean;
    viewSlide: (id: number) => any;
}

export interface SlideProps {
    slide: SlideModel;
}
