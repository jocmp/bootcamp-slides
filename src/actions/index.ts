import { SlideshowModel } from '../Models'
import { History } from 'history';

export const nextSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const nextIndex: number = currentIndex < totalLength - 1 ? currentIndex + 1 : currentIndex;
    return handleSlide(id, nextIndex, history);
}

export const previousSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const previousIndex: number = currentIndex > 0 ? currentIndex - 1 : currentIndex;
    return handleSlide(id, previousIndex, history);
}

const handleSlide = (id: string, index: number, history: History) => {
    history.push(`/slideshows/${id}/slides/${index}`);
    return viewSlide(index);
}

export const viewSlide = (index: number) => {
    return {
        type: 'VIEW_SLIDE',
        index
    };
}

const apiData = (index: number): SlideshowModel => {
    const shows: SlideshowModel[] = [
        {
            id: 1,
            title: "Exceptional Presentation",
            slides: [
                {
                    title: "Title here",
                    slide_type: "title"
                },
                {
                    title: "Sensational Informational",
                    slide_type: "simple",
                    content: ["Inspirational insatiable"]
                }
            ]
        },
        {
            id: 2,
            title: "Touch Fuzzy, Get Dizzy",
            slides: [
                {
                    title: "Yoshi's Island",
                    slide_type: "title",
                },
                {
                    title: "Yoshi's Island",
                    slide_type: "simple",
                    content: ['Imperically the best Super Mario game... \ntest']
                },
                {
                    title: "Why buy a SNES?",
                    slide_type: "simple",
                    content: ["Yoshi's Island. No question."]
                }
            ]
        }
    ]
    return shows[index - 1];
};

export const loadSlideshow = (slideshow: SlideshowModel, currentIndex: number) => {
    return {
        type: 'LOAD_SLIDESHOW',
        slideshow,
        currentIndex
    }
};

export const fetchSlideshow = (id: number, index: number = 0) => (dispatch: any, getState: any) => {
    dispatch(loadSlideshow(apiData(id), index));
};