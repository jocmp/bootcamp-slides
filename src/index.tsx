import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './components/App';
import './index.css';
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

export const slideshow = (state: any, action: any) => {
  switch(action.type) {
    case 'LOAD_SLIDESHOW':
      return {
        slideshow: action.slideshow
      }
    default:
      return {};
  }
}

export const store = createStore(slideshow, applyMiddleware(thunk));
store.subscribe(() => console.log(store.getState()))

ReactDOM.render(
  <App store={store} />,
  document.getElementById('root') as HTMLElement
);
