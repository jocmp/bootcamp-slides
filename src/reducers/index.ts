import { combineReducers } from 'redux'
import { AppState } from '../Models'

const viewedIndices = (state: number[] = [], action: any) => {
    switch (action.type) {
        case 'VIEW_SLIDE':
            return Array.from(new Set([...state, action.index]));
        default:
            return state;
    }
};

const initialSlideshow: any = { title: "", id: -1, slides: [] };

const slideshow = (state: AppState, action: any) => {
    switch (action.type) {
        case 'LOAD_SLIDESHOW':
            return action.slideshow;
        case 'LOAD_SLIDESHOW_ERROR':
            return initialSlideshow;
        case 'VIEW_SLIDE':
        default:
            if (!!state) {
                return state;
            }
            return initialSlideshow;
    }
}

const fetchSlideshow = (state: AppState, action: any) => {
    switch (action.type) {
        case 'FETCH_SLIDESHOW':
            return action.loading;
        default:
            return false;
    }
}

const error = (state: AppState, action: any) => {
    switch (action.type) {
        case 'LOAD_SLIDESHOW_ERROR':
            return action.error;
        default:
            return '';
    }
}

const rootReducer = combineReducers({
    slideshow,
    viewedIndices,
    fetchSlideshow,
    error
});

export default rootReducer;