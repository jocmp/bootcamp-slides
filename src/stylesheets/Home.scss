@import "./_clickable";

.home-container {
    margin-top: 1%;
    margin-left: 20px; 
    display: flex;
    height: 60vh;
    width: 100%;
    align-items: center;
    justify-content: center; 

    label {
        margin-right: 16px;
    } 
    
    .slide-button {
        @extend .clickable;
        margin-left: 16px;
    }
}