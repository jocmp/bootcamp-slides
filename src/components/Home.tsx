import * as React from 'react';
import '../stylesheets/Home.scss';
import { HomeProps } from '../Props';
import { LinkFactory } from '../Models';
import { Link } from 'react-router-dom';

const logo = require('../assets/wordmark.png');

class Home extends React.Component<HomeProps, { selectedSlideshowId: string }> {

    constructor(props: HomeProps) {
        super(props);
        this.state = { selectedSlideshowId: '' };
        console.log(this.props.slideshow);
    }

    handleChange = (event: any) => {
        this.props.clearError();
        this.setState({ selectedSlideshowId: event.target.value })
        if (!isNaN(Number(event.target.value))) {
            const searchValue = event.target.value;
            this.props.searchSlideshows(Number(searchValue))
        }
    }

    render() {
        return (
            <div className="home-container">
                <div className="search-header">
                    <img src={logo} className="search-logo" alt="logo" />
                    <input placeholder="Search slideshow IDs" onChange={this.handleChange} value={this.state.selectedSlideshowId} type="text" />
                </div>
                <Link className={this.props.slideshow.id > 0 ? "slide-button" : "button-invisible"}
                    to={LinkFactory.createSlideshowLink(this.props.slideshow.id)}>
                    {this.props.slideshow.title}
                </Link>
                {this.props.error.length > 0 && this.state.selectedSlideshowId &&
                    <span className="slide-error">No Slideshows Found.</span>
                }
            </div>
        )
    }
}

export default Home;