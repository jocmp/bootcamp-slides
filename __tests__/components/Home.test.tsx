import * as React from 'react';
import Home from '../../src/components/Home';
import { SlideshowModel } from '../../src/Models';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('Home', () => {

    const defaultSlideshow: SlideshowModel = { title: 'Slideshow', id: 1, slides: [] };

    const getWrapper = (slideshow: SlideshowModel, loading: boolean = false, error: string = '',
        searcher: (event: any) => void = jest.fn()) => {
        return shallow(
            <Home searchSlideshows={searcher} loading={loading} error={error} slideshow={slideshow} />
        );
    };

    it('renders', () => {
        expect(toJson(getWrapper(defaultSlideshow))).toMatchSnapshot();
    });

    it('shows errors when present', () => {
        const message = 'No Slideshows Found.';
        expect(getWrapper(defaultSlideshow, false, message).text()).toContain(message);
    });

    it('shows a button for the slideshow', () => {
        expect(getWrapper(defaultSlideshow).find('.slide-button')).toHaveLength(1);
    });

    it('calls search on text change', () => {
        const searcher = jest.fn();
        const wrapper = getWrapper(defaultSlideshow, false, '', searcher);
        wrapper.find('input').simulate('keyDown', { keyCode: 49 });
        setTimeout(() =>
            expect(searcher).toHaveBeenCalled(),
            200);
    });
});