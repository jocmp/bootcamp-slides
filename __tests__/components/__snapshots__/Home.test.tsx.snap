// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<div
  className="home-container"
>
  <label>
    Slideshow ID
  </label>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <Link
    className="slide-button"
    replace={false}
    to="/slideshows/1/slides/0"
  >
    Slideshow
  </Link>
</div>
`;
